### PR TITLE
Migrate basic aer provider to new versioned interface

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit migrates the basic aer provider to the v2 interface. This
was originally included in #5086 but had to be removed because of a
potential backwards compatibility issue with aqua when using basic aer
as the provider (aqua 0.7.x explicity checks for v1 interface backends).

### Details and comments

### CODE REVIEW

**Comment by mtreinish** [comment]
This depends on #5086 and will fail CI on the tutorials job until Qiskit/qiskit-aqua#1270 merges and is released. This means this is blocked until after 0.16 is released. I will rebase it after #5086 merges to reduce the diff.

**Comment by mtreinish** [comment]
This is going to be blocked until the next aqua release. The quantum instance in aqua is accessing private attributes which is broken by the changes I make here. So we need to update the quantum instance in aqua to unblock the tutorials job (in addition to fixing the code migrated from aqua which also has this bug and is blocking this too)

File: releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml
Comment (ajavadia) [inline releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml:35]: incomplete sentence here

File: releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml
Comment (ajavadia) [inline releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml:31]:
```suggestion
    from BasicAer backends are now synchronous instances of
```

File: releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml
Comment (ajavadia) [inline releasenotes/notes/basicaer-new-provider-ea7cf756df231c2b.yaml:9]:
```suggestion
    :class:`~qiskit.circuit.QuantumCircuit` objects) as its input.
```

File: qiskit/providers/basicaer/qasm_simulator.py
Comment (ajavadia) [inline qiskit/providers/basicaer/qasm_simulator.py:158]: I think `allow_sample_measuring` should be True by default otherwise it will be very slow (1000x slower for a 1000-shot simulation).

Comment (mtreinish) [reply qiskit/providers/basicaer/qasm_simulator.py:158]: I was just mirroring what was the default before. I'll change it here and add an upgrade release note about the change

```suggestion
                       allow_sample_measuring=True, seed_simulator=None,
```

Comment (mtreinish) [reply qiskit/providers/basicaer/qasm_simulator.py]: I actually think we'll be ok without a release note for this. Looking at the code for checking the option: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/providers/basicaer/qasm_simulator.py#L340-L378 it would default to ``True`` if nothing was set if the circuit is capable of using sample measuring. So I think putting either value here doesn't matter too much as it doesn't really change the behavior.

**Review by ajavadia — APPROVED** [review]

**Pull request merged**
